examples/go/tagging/handler/handler.go

```diff
@@ -31,140 +32,145 @@ type vcConfig struct {
+type incoming struct {
+	Data types.Event `json:"data,omitempty"`
 }
 
+var (
+	client *vsClient.Client
+	ctx    = context.Background()
+	cfg    vcConfig
+	sigCh  = make(chan os.Signal)
+)
+
 // Handle a function invocation
 func Handle(req handler.Request) (handler.Response, error) {
-	ctx := context.Background()
-
-	// Retrieve vCenter and tag info from the config file.
-	cfg, err := loadTomlCfg()
-	if err != nil {
-		log.Error(err)
-		return
+	// Connect to vSphere govmomi API if not already done. Also, load the required
+	// vcconfig.
+	if client == nil {
+		err := vsConnect()
+		if err != nil {
+			log.Fatalf("connect to vSphere failed: %v", err)
+			return handler.Response{
+				Body:       nil,
+				StatusCode: http.StatusBadRequest,
+			}, fmt.Errorf("connect to vSphere failed: %w", err)
+		}
 	}
 
-	// Incoming request object.
-	var event vmEvent
-	err := json.Unmarshal(req.Body, &event)
+	// Set up os signal handling to log out of vSphere.
+	signal.Notify(sigCh, syscall.SIGTERM, os.Interrupt)
+	go handleSignal(ctx, sigCh)
+
+	// Retrieve the Managed Object Reference from the event.
+	moRef, err := parseEventMoRef(req.Body)
 	if err != nil {
+		log.Fatalf("retrieve managed reference object failed: %v", err)
 		return handler.Response{
 			Body:       nil,
 			StatusCode: http.StatusBadRequest,
 		}, err
 	}
 
-	if event.Event.Vm.Vm.Value == "" || event.Event.Vm.Vm.Type == "" {
+	log.Println("tag VM moRef with tagID (URN)")
+	err = client.MoTag(ctx, moRef, cfg.Tag.Urn)
+	if err != nil {
+		log.Fatalf("tagging managed reference object failed: %v", err)
 		return handler.Response{
 			Body:       nil,
 			StatusCode: http.StatusBadRequest,
-		}, errors.New("managed Reference Object information must not be empty")
+		}, err
 	}
 
-	// Fill information in the request into a govmomi type.
-	var moRef types.ManagedObjectReference
-	moRef.Value = event.Event.Vm.Vm.Value
-	moRef.Type = event.Event.Vm.Vm.Type
-
-	// Connect to vcenter with govmomi client
-	log.Println("connect to vCenter")
-	clients := connect(ctx, vcc.VCenter)
-	log.Println("connection successful")
-
-	// tag the VM moRef with tagID (URN)
-	clients.tagVM(ctx, moRef, vcc.Tag.Urn)
-
 	// Respond with a message that tagging of vm was complete
 	message := fmt.Sprintf(
 		"%v was tagged with %v",
 		moRef.Value,
-		vcc.Tag.Urn,
+		cfg.Tag.Urn,
 	)
 	log.Println(message)
 
-	err = clients.Govmomi.Logout(ctx)
-	if err != nil {
-		log.Printf("unable to log out of govmomi due to %v", err)
-	}
-	err = clients.Rest.Logout(ctx)
-	if err != nil {
-		log.Printf("unable to log out of rest due to %v", err)
-	}
-
 	return handler.Response{
 		Body:       []byte(message),
 		StatusCode: http.StatusOK,
 	}, nil
 }
 
-func loadTomlCfg() (vcConfig, error) {
-	secretPath := "/var/openfaas/secrets/vcconfig"
-	secret, err := toml.LoadFile(secretPath)
+// vsConnect connects to vSphere govmomi API using information from vcconfig.toml.
+func vsConnect() error {
+	var (
+		err        error
```

Comment:
Yes, I believe so. Declaring error ahead of time is needed so I can avoid short variable declaration, `:=` the first time err is assigned in 
```go
cfg, err = loadTomlCfg(configPath)
```

I have found that strange things happen when we try to use short variable declaration for a multi-variable return function and one of the variables to be assigned is a global variable. Here's where I show this unexpected behavior: https://play.golang.org/p/W1wmkpAAqTN

You would think that if we did `cfg, err := loadTomlCfg()`, that Go would realize `cfg` was a global variable and `err` was new. But it doesn't. It's as if it creates a new, shadow `cfg` variable. Because of this, I had to declare `err` ahead of time, and just use the `=` assignment operator.

Reply:
OK, gotcha. My linter did not show this as shadowed. Either way, this can be solved with:

1) Using different variable for the cfg and set the global in a two-step approach
2) Set cfg to nil (or default value) because the shadowed variable would not affect the global anyways if it's an error returned (leaving the global in default/nil state - which is what you want anyways)

I'll let you decide.

Reply:
(also just saw discussion below on global scope of this var ;)

Reply:
I ended up keeping `var err error` for a different reason: sync.Once does not allow for functions with returns. however, I still wanted to capture errors, so I declared the variable outside `once.Do()`.